crates/router/src/db/capture.rs

```diff
@@ -0,0 +1,403 @@
+use diesel_models::payment_attempt::PaymentAttemptUpdate;
+use error_stack::ResultExt;
+
+use super::MockDb;
+use crate::{
+    core::errors::{self, CustomResult},
+    db::payment_attempt::PaymentAttemptInterface,
+    types::storage::{self as types, enums, errors as storage_errors},
+    utils,
+};
+
+#[async_trait::async_trait]
+pub trait CaptureInterface {
+    async fn insert_capture(
+        &self,
+        capture: types::CaptureNew,
+        parent_attempt: types::PaymentAttempt,
+        storage_scheme: enums::MerchantStorageScheme,
+    ) -> CustomResult<(types::Capture, types::PaymentAttempt), errors::StorageError>;
+
+    async fn find_all_captures_by_authorized_attempt_id(
+        &self,
+        authorized_attempt_id: &str,
+        storage_scheme: enums::MerchantStorageScheme,
+    ) -> CustomResult<Vec<types::Capture>, errors::StorageError>;
+
+    async fn find_all_charged_captures_by_authorized_attempt_id(
+        &self,
+        authorized_attempt_id: &str,
+        storage_scheme: enums::MerchantStorageScheme,
+    ) -> CustomResult<Vec<types::Capture>, errors::StorageError>;
+
+    async fn update_capture_and_attempt_with_capture_id(
+        &self,
+        this: types::Capture,
+        parent_attempt: types::PaymentAttempt,
+        intent: &types::PaymentIntent,
+        capture: types::CaptureUpdate,
+        storage_scheme: enums::MerchantStorageScheme,
+    ) -> CustomResult<(types::Capture, types::PaymentAttempt), errors::StorageError>;
+
+    fn validate_attempt_and_capture(
+        &self,
+        parent_attempt: &types::PaymentAttempt,
+        authorized_attempt_id: &String,
+    ) -> CustomResult<(), errors::StorageError> {
+        utils::when(&parent_attempt.attempt_id != authorized_attempt_id, || {
+            Err(
+                errors::StorageError::DatabaseError(storage_errors::DatabaseError::Others.into())
+                    .into(),
+            )
+            .attach_printable("authorized_attempt_id of capture did not match parent attempt_id")
+        })
+    }
+}
+
+#[cfg(feature = "kv_store")]
```

Comment:
I see that both the implementations are similar, is something yet to be done for the kv feature? If so then create an issue and link it here.